File: internal/projects/shellrun.go
package projects

import (
	"bytes"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
)

// RunScriptInDir runs a shell one-liner with working directory dir (e.g. before removing a worktree).
// Stdout and stderr are captured for error messages.
func RunScriptInDir(dir, script string) error {
	if runtime.GOOS == "windows" {
		return fmt.Errorf("los scripts de proyecto no están soportados en Windows")
	}
	line := strings.TrimSpace(script)
	if line == "" {
		return nil
	}
	line = ExpandScriptPlaceholders(line, dir)
	absDir, err := filepath.Abs(dir)
	if err != nil {
		return err
	}
	shell := os.Getenv("SHELL")
	if shell == "" {
		shell = "/bin/sh"
	}
	shellPath, err := exec.LookPath(shell)
	if err != nil {
		shellPath = "/bin/sh"
	}
	cmd := exec.Command(shellPath, "-lc", line)
	cmd.Dir = absDir
	cmd.Env = os.Environ()
	var out bytes.Buffer
	cmd.Stdout = &out
	cmd.Stderr = &out
	if err := cmd.Run(); err != nil {
		tail := strings.TrimSpace(out.String())
		if tail != "" {
			return fmt.Errorf("%w: %s", err, tail)
		}
		return err
	}
	return nil
}

// RunScriptCapture runs the same shell one-liner as RunScriptInDir and returns combined stdout+stderr.
// If the command fails, the captured output is still returned together with the error.
func RunScriptCapture(dir, script string) (string, error) {
	if runtime.GOOS == "windows" {
		return "", fmt.Errorf("los scripts de proyecto no están soportados en Windows")
	}
	line := strings.TrimSpace(script)
	if line == "" {
		return "", nil
	}
	absDir, err := filepath.Abs(dir)
	if err != nil {
		return "", err
	}
	line = ExpandScriptPlaceholders(line, absDir)
	shell := os.Getenv("SHELL")
	if shell == "" {
		shell = "/bin/sh"
	}
	shellPath, err := exec.LookPath(shell)
	if err != nil {
		shellPath = "/bin/sh"
	}
	cmd := exec.Command(shellPath, "-lc", line)
	cmd.Dir = absDir
	cmd.Env = os.Environ()
	var out bytes.Buffer
	cmd.Stdout = &out
	cmd.Stderr = &out
	err = cmd.Run()
	return out.String(), err
}

File: src/projects/shellrun.ts
import { spawn } from "node:child_process";
import path from "node:path";
import { expandScriptPlaceholders } from "./projectConfig.js";

function shellPath(): string {
  return process.env.SHELL?.trim() || "/bin/sh";
}

export async function runScriptInDir(dir: string, script: string): Promise<void> {
  if (process.platform === "win32") {
    throw new Error("los scripts de proyecto no están soportados en Windows");
  }
  let line = script.trim();
  if (!line) {
    return;
  }
  line = expandScriptPlaceholders(line, dir);
  const absDir = path.resolve(dir);
  const sh = shellPath();
  await new Promise<void>((resolve, reject) => {
    const chunks: Buffer[] = [];
    const child = spawn(sh, ["-lc", line], {
      cwd: absDir,
      env: process.env,
      stdio: ["ignore", "pipe", "pipe"],
    });
    child.stdout?.on("data", (d: Buffer) => chunks.push(d));
    child.stderr?.on("data", (d: Buffer) => chunks.push(d));
    child.on("error", reject);
    child.on("close", (code) => {
      const s = Buffer.concat(chunks).toString("utf8").trim();
      if (code === 0) {
        resolve();
      } else {
        reject(s ? new Error(`exit ${code}: ${s}`) : new Error(`script failed (${code})`));
      }
    });
  });
}

export async function runScriptCapture(
  dir: string,
  script: string
): Promise<string> {
  if (process.platform === "win32") {
    throw new Error("los scripts de proyecto no están soportados en Windows");
  }
  let line = script.trim();
  if (!line) {
    return "";
  }
  const absDir = path.resolve(dir);
  line = expandScriptPlaceholders(line, absDir);
  const sh = shellPath();
  return new Promise((resolve, reject) => {
    const chunks: Buffer[] = [];
    const child = spawn(sh, ["-lc", line], {
      cwd: absDir,
      env: process.env,
      stdio: ["ignore", "pipe", "pipe"],
    });
    child.stdout?.on("data", (d: Buffer) => chunks.push(d));
    child.stderr?.on("data", (d: Buffer) => chunks.push(d));
    child.on("error", reject);
    child.on("close", (code) => {
      const s = Buffer.concat(chunks).toString("utf8");
      if (code === 0) {
        resolve(s);
      } else {
        reject(new Error(s.trim() ? s.trim() : `exit ${code}`));
      }
    });
  });
}
